x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,507 @@
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+			}
+			p.logger.Infof("Message visibility timeout updated to %v", visibilityTimeout)
+		}
+	}
+}
+
+func (p *Input) changeVisibilityTimeout(queueURL string, visibilityTimeout int64, svc *sqs.Client, receiptHandle *string) error {
+	req := svc.ChangeMessageVisibilityRequest(&sqs.ChangeMessageVisibilityInput{
+		QueueUrl:          &queueURL,
+		VisibilityTimeout: &visibilityTimeout,
+		ReceiptHandle:     receiptHandle,
+	})
+	_, err := req.Send(p.context)
+	return err
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// handle message
+func handleSQSMessage(m sqs.Message) ([]s3Info, error) {
+	msg := sqsMessage{}
+	err := json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "json unmarshal sqs message body failed")
+	}
+
+	var s3Infos []s3Info
+	for _, record := range msg.Records {
+		if record.EventSource == "aws:s3" && record.EventName == "ObjectCreated:Put" {
+			s3Infos = append(s3Infos, s3Info{
+				region: record.AwsRegion,
+				name:   record.S3.bucket.Name,
+				key:    record.S3.object.Key,
+				arn:    record.S3.bucket.Arn,
+			})
+		}
+	}
+	return s3Infos, nil
+}
+
+func (p *Input) handleS3Objects(svc s3iface.ClientAPI, s3Infos []s3Info, errC chan error) error {
+	s3Context := &s3Context{
+		refs: 1,
+		errC: errC,
+	}
+
+	for _, s3Info := range s3Infos {
+		objectHash := s3ObjectHash(s3Info)
+
+		// read from s3 object
+		reader, err := newS3BucketReader(svc, s3Info, p.context)
+		if err != nil {
+			return errors.Wrap(err, "newS3BucketReader failed")
+		}
+
+		offset := 0
+		for {
+			log, err := reader.ReadString('\n')
+			if log == "" {
+				break
+			}
+
+			if err != nil {
+				if err == io.EOF {
+					// create event for last line
+					offset += len([]byte(log))
+					event := createEvent(log, offset, s3Info, objectHash, s3Context)
+					err = p.forwardEvent(event)
+					if err != nil {
+						err = errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+						s3Context.Fail(err)
+						return err
+					}
+					return nil
+				}
+				return errors.Wrapf(err, "ReadString failed for %v", s3Info.key)
+			}
+
+			// create event per log line
+			offset += len([]byte(log))
+			event := createEvent(log, offset, s3Info, objectHash, s3Context)
+			err = p.forwardEvent(event)
+			if err != nil {
+				err = errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+				s3Context.Fail(err)
+				return err
+			}
+		}
+	}
+
+	s3Context.done()
```

Comment:
better use `defer` at the top. This function can return on errors, in which case you would miss the `done` setting the count to 0.